src/tools/jsondocck/src/cache.rs

```diff
@@ -1,77 +1,50 @@
 use serde_json::Value;
 use std::collections::HashMap;
-use std::io;
 use std::path::{Path, PathBuf};
 
 use fs_err as fs;
 
+#[derive(Debug)]
+pub struct FileInfo {
+    pub path: PathBuf,
+    pub content: String,
+}
+
 #[derive(Debug)]
 pub struct Cache {
-    root: PathBuf,
-    files: HashMap<PathBuf, String>,
-    values: HashMap<PathBuf, Value>,
+    file: FileInfo,
+    value: Option<Value>,
     pub variables: HashMap<String, Value>,
-    last_path: Option<PathBuf>,
 }
 
 impl Cache {
     /// Create a new cache, used to read files only once and otherwise store their contents.
-    pub fn new(doc_dir: &str) -> Cache {
+    pub fn new(config: &Config) -> Cache {
+        let root = Path::new(&config.doc_dir);
+        let filename = Path::new(&config.template).file_stem().unwrap();
+        let file_path = root.join(&format!("{}.json", filename.to_str().unwrap()));
         Cache {
-            root: Path::new(doc_dir).to_owned(),
-            files: HashMap::new(),
-            values: HashMap::new(),
+            file: FileInfo {
+                path: file_path.to_owned(),
+                content: fs::read_to_string(&file_path).expect("failed to read JSON file"),
+            },
+            value: None,
             variables: HashMap::new(),
-            last_path: None,
-        }
-    }
-
-    fn resolve_path(&mut self, path: &String) -> PathBuf {
-        if path != "-" {
-            let resolve = self.root.join(path);
-            self.last_path = Some(resolve.clone());
-            resolve
-        } else {
-            self.last_path
-                .as_ref()
-                // FIXME: Point to a line number
-                .expect("No last path set. Make sure to specify a full path before using `-`")
-                .clone()
         }
     }
 
-    fn read_file(&mut self, path: PathBuf) -> Result<String, io::Error> {
-        if let Some(f) = self.files.get(&path) {
-            return Ok(f.clone());
-        }
-
-        let file = fs::read_to_string(&path)?;
-
-        self.files.insert(path, file.clone());
-
-        Ok(file)
-    }
-
-    /// Get the text from a file. If called multiple times, the file will only be read once
-    pub fn get_file(&mut self, path: &String) -> Result<String, io::Error> {
-        let path = self.resolve_path(path);
-        self.read_file(path)
-    }
-
     /// Parse the JSON from a file. If called multiple times, the file will only be read once.
-    pub fn get_value(&mut self, path: &String) -> Result<Value, CkError> {
-        let path = self.resolve_path(path);
-
-        if let Some(v) = self.values.get(&path) {
-            return Ok(v.clone());
+    pub fn get_value(&mut self) -> Result<&Value, CkError> {
+        if let Some(ref v) = self.value {
+            return Ok(v);
         }
 
-        let content = self.read_file(path.clone())?;
-        let val = serde_json::from_str::<Value>(&content)?;
+        let val = serde_json::from_str::<Value>(&self.file.content)?;
```

Comment:
Why do you do this here, instead of deserializing the file content in `Cache::new`?

Reply:
Very good point!